### PR TITLE
Add some sanity checks for audio management

### PR DESCRIPTION
**Pull request by kOchirasu**
Sorry for so many PRs, but it should be good now. Not really fixing anything here just adding sanity checks.

Some takeaways:
- Do not call `Free` on the **active** device (Calling `Free` on an inactive device is fine). Active is defined as the device assigned by `Bass.ChannelSetDevice`.
- If a device was disabled and enabled again, it needs to be initialized again even though it will look to be already initialized from flags.
- If you switch from device A to B, then back to A again, you actually do not need to reinitialize A. Setting CurrentDevice to A is sufficient. (We don't differentiate this in the code and just re-init, but it seems to work fine)

### CODE REVIEW

**Comment by 2yangk23** [comment]
For some reason tests failed because I moved the Free call into AudioManager. Looking in to it...

**Comment by 2yangk23** [comment]
Don't know why this keeps failing. Passes locally every time with and without audio devices enabled.

File: osu.Framework/Threading/AudioThread.cs [hunk 1]
```diff
@@ -75,17 +75,16 @@ protected override void PerformExit()
...
-                    manager.Dispose();
+                managersCopy = new List<AudioManager>(managers);
                 managers.Clear();
             }
 
-            // Safety net to ensure we have freed all devices before exiting.
-            // This is mainly required for device-lost scenarios.
-            // See https://github.com/ppy/osu-framework/pull/3378 for further discussion.
-            while (Bass.Free()) { }
+            foreach (var manager in managersCopy)
```
Comment (peppy) [inline osu.Framework/Threading/AudioThread.cs]: what's the reason for making a copy here?

Comment (2yangk23) [reply osu.Framework/Threading/AudioThread.cs]: It is a deadlock risk (during testing) since Dispose is called synchronously rather than schedules on AudioThread.

case:
1. AudioThread.Dispose is called
2. AudioThread acquires lock on managers
3. AudioThread calls dispose for all manager in managers
4. manager.Dipose is called
5. manager calls AudioThread.UnregisterManager
6. manager tries to acquire lock on managers (deadlock)

Comment (peppy) [reply osu.Framework/Threading/AudioThread.cs]: I see, due to the schedule of the dispose. We are 100% sure we do want to schedule the disposal calls still?

Comment (2yangk23) [reply osu.Framework/Threading/AudioThread.cs]: I don't know that scheduling the dispose call makes sense because it gets scheduled and AudioThread is immediately killed, so it will never be run. I remember that you had mentioned that it's scheduled because we need it to run on the AudioThread, so possibly what needs to happen is something like:

1. AudioThread.Dispose is called
2. AudioThread calls dispose for all managers, and unregisters them
3. AudioThread waits until all scheduled work is complete (how can this be guaranteed?)
4. AudioThread.Dispose completes

Comment (peppy) [reply osu.Framework/Threading/AudioThread.cs]: But this code here is guaranteed to be on the audio thread, right? So we can potentially call the non-queueing method I think?

Trying to understand where `manager.Dispose` is being called from, is it a finalizer or otherwise?

Comment (2yangk23) [reply osu.Framework/Threading/AudioThread.cs]: `manager.Dispose()` is normally called by `AudioThread.Dispose` and it schedules `manager.Dispose(bool)` in the AudioThread scheduler. I'm not sure where `AudioThread.Dispose` is called from.

During these I overrode `manager.Dispose()` in `AudioManagerWithDeviceLoss` to directly call `manager.Dispose(bool)` otherwise `Bass.Free` is never called and causes tests to interfere with each other.

Comment (peppy) [reply osu.Framework/Threading/AudioThread.cs]: Hmm, this is all very confusing and hard to follow. I'll have to debug and figure out the best path forward, since we really shouldn't be able to daedlock at this point, all going well.

Comment (2yangk23) [reply osu.Framework/Threading/AudioThread.cs]: With the current implementation, the deadlock should not happen normally (it's only a risk during testing)

Call stack would be something like this during testing:
```
AudioThread.Dispose() (lock)
>manager.Dispose()
>>manager.Dispose(bool)
>>>AudioThread.UnregisterManager() (deadlock)
```
what happens otherwise is:
```
AudioThread.Dispose() (lock)
>manager.Dispose()
>>schedule(manager.Dispose(bool))
(lock released)
[scheduler] manager.Dispose(bool) (lock)
```

**Comment by peppy** [comment]
Looks like this failed on CI :(.
